romana/kubernetes/listener.go

```diff
@@ -0,0 +1,364 @@
+//	}
+//}
+
+// Run implements the main loop, reconnecting as needed.
+//func RunListener0(rootURL string, kubeURL string) {
+//	l := kubeListener{kubeURL: kubeURL, restClient: common.NewRestClient(common.GetDefaultRestClientConfig(rootURL))}
+//	l.Run()
+//}
+//
+//func (l kubeListener) Run0() {
+//	for {
+//		ch := make(chan []byte)
+//		go listen(ch, l.kubernetesURL)
+//		for {
+//			chunk, ok := <-ch
+//			log.Println(ok)
+//			if chunk != "" {
+//				log.Printf("Read chunk %v\n%s\n------------", ok, string(chunk))
+//				err = processChunk(chunk)
+//				if err != nil {
+//					// TODO is there any other way to handle this?
+//					log.Printf("Error processing chunk %s: %v", string(chunk), err)
+//				}
+//			}
+//			if !ok {
+//				break
+//			}
+//		}
+//	}
+//}
+
+/////////////////////////////////////////////////////////////
+
+type networkPolicyAction int
+
+const (
+	networkPolicyActionDelete networkPolicyAction = iota
+	networkPolicyActionAdd
+	networkPolicyActionModify
+)
+
+type kubeListener struct {
+	config           common.ServiceConfig
+	restClient       common.RestClient
+	kubeUrl          string
+	urlPrefix        string
+	segmentLabelName string
+}
+
+// Routes returns various routes used in the service.
+func (l *kubeListener) Routes() common.Routes {
+	routes := common.Routes{}
+	return routes
+}
+
+// Name implements method of Service interface.
+func (l *kubeListener) Name() string {
+	return "kubernetes-listener"
+}
+
+// SetConfig implements SetConfig function of the Service interface.
+func (l *kubeListener) SetConfig(config common.ServiceConfig) error {
+	m := config.ServiceSpecific
+	l.kubeUrl = m["kubernetes_url"].(string)
+	if l.kubeUrl == "" {
+		return errors.New("kubernetes_url required.")
+	}
+	l.urlPrefix = m["url_prefix"].(string)
+	if l.urlPrefix == "" {
+		return errors.New("url_prefix required.")
+	}
+	l.segmentLabelName = m["segment_label_name"].(string)
+	if l.segmentLabelName == "" {
+		return errors.New("segment_label_name required.")
+	}
+
+	return nil
+
+}
+
+// Run configures and runs listener service.
+func Run(rootServiceURL string, cred *common.Credential) (*common.RestServiceInfo, error) {
+	clientConfig := common.GetDefaultRestClientConfig(rootServiceURL)
+	clientConfig.Credential = cred
+	client, err := common.NewRestClient(clientConfig)
+	if err != nil {
+		return nil, err
+	}
+	kubeListener := &kubeListener{}
+	config, err := client.GetServiceConfig(kubeListener)
+	if err != nil {
+		return nil, err
+	}
+	return common.InitializeService(kubeListener, *config)
+}
+
+// translateNetworkPolicy translates a Kubernetes policy into
+// Romana policy (see common.Policy) with the following rules:
+// 1. Kubernetes Namespace corresponds to Romana Tenant
+// 2. If Romana Tenant does not exist it is an error (a tenant should
+//    automatically be created when a namespace is added, however)
+func (l *kubeListener) translateNetworkPolicy(kubePolicy *KubeObject) (common.Policy, error) {
```

Comment:
This function is too long. If it can be refactored into smaller functions, that would be great. I would even be happy if it would at least have some clearly documented code blocks.